<launch>
    <arg name="localization" default="false"/>
    <arg if="$(arg localization)" name="rtabmap_args" default=""/>
    <arg unless="$(arg localization)" name="rtabmap_args" default="-d"/>
    <arg name="rtabmap_perception_delay"/>

    <node pkg="nodelet" type="nodelet" name="rgbd_sync" args="standalone rtabmap_sync/rgbd_sync" launch-prefix="bash -c 'sleep $(arg rtabmap_perception_delay); $0 $@' ">
        <remap from="rgb/image" to="/camera_3d/color/image_raw"/>
        <remap from="depth/image" to="/camera_3d/aligned_depth_to_color/image_raw"/>
        <remap from="rgb/camera_info" to="/camera_3d/aligned_depth_to_color/camera_info"/>

        <param name="approx_sync" value="true"/>
    </node>
    <node pkg="hbba_lite" type="on_off_hbba_filter_node" name="rtabmap_filter" launch-prefix="bash -c 'sleep $(arg rtabmap_perception_delay); $0 $@' ">
        <remap from="in" to="rgbd_image"/>
        <remap from="out" to="rgbd_image_filtered"/>
        <remap from="filter_state" to="rtabmap/filter_state"/>
    </node>
    <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" args="$(arg rtabmap_args)" launch-prefix="bash -c 'sleep $(arg rtabmap_perception_delay); $0 $@' ">
        <param name="odom_frame_id" type="string" value="base_link"/>
        <param name="subscribe_depth" type="bool" value="false"/>
        <param name="subscribe_rgbd" type="bool" value="true"/>
        <param name="subscribe_rgb" type="bool" value="false"/>

        <param name="queue_size" type="int" value="10"/>

        <param name="map_always_update" type="bool" value="true"/>
        <param name="map_filter_radius" type="double" value="0.1"/>
        <param name="map_filter_angle" type="double" value="10"/>
        <param name="gen_depth" type="bool" value="false"/>

        <param name="RGBD/AngularUpdate" type="string" value="0"/>
        <param name="RGBD/LinearUpdate" type="string" value="0"/>
        <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>

        <!-- Occupancy Grid -->
        <param name="Grid/NormalsSegmentation" value="false"/>
        <param name="Grid/MaxObstacleHeight" value="0.1"/>
        <param name="Grid/MinGroundHeight" value="-0.1"/>
        <param name="Grid/MaxGroundHeight" value="0.01"/>
        <param name="Grid/3D" value="true"/>
        <param name="Grid/RayTracing" value="true"/>

        <!-- localization mode -->
        <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
        <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
        <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>

        <remap from="rgbd_image" to="rgbd_image_filtered"/>
    </node>

</launch>
